ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionTimeGranularityOptimizer.java

```diff
@@ -0,0 +1,353 @@
+        } else {
+          descs.add(new ExprNodeColumnDesc(col.getTypeInfo(), ReduceField.VALUE.toString()+"."+colName, null, false));
+        }
+      }
+      RowSchema selRS = new RowSchema(granularitySelOp.getSchema());
+      SelectDesc selConf = new SelectDesc(descs, colNames);
+      SelectOperator backtrackSelOp = (SelectOperator) OperatorFactory.getAndMakeChild(
+          selConf, selRS, rsOp);
+
+      // Link backtrack SelectOp to FileSinkOp
+      fsOp.getParentOperators().clear();
+      fsOp.getParentOperators().add(backtrackSelOp);
+      backtrackSelOp.getChildOperators().add(fsOp);
+
+      // Update file sink descriptor
+      fsOp.getConf().setDpSortState(FileSinkDesc.DPSortState.PARTITION_SORTED);
+      fsOp.getConf().setPartitionCols(rsOp.getConf().getPartitionCols());
+      ColumnInfo ci = new ColumnInfo(granularitySelOp.getSchema().getSignature().get(
+              granularitySelOp.getSchema().getSignature().size() - 1)); // granularity column
+      fsOp.getSchema().getSignature().add(ci);
+
+      LOG.info("Inserted " + granularitySelOp.getOperatorId() + ", " + rsOp.getOperatorId() + " and "
+          + backtrackSelOp.getOperatorId() + " as parent of " + fsOp.getOperatorId()
+          + " and child of " + fsParent.getOperatorId());
+
+      parseCtx.setReduceSinkAddedBySortedDynPartition(true);
+      return null;
+    }
+
+    private Operator<? extends OperatorDesc> getGranularitySelOp(
+            Operator<? extends OperatorDesc> fsParent) throws SemanticException {
+      ArrayList<ColumnInfo> parentCols = Lists.newArrayList(fsParent.getSchema().getSignature());
+      ArrayList<ExprNodeDesc> descs = Lists.newArrayList();
+      List<String> colNames = Lists.newArrayList();
+      int timestampPos = -1;
+      for (int i = 0; i < parentCols.size(); i++) {
+        ColumnInfo ci = parentCols.get(i);
+        ExprNodeColumnDesc columnDesc = new ExprNodeColumnDesc(ci);
+        descs.add(columnDesc);
+        colNames.add(columnDesc.getExprString());
+        if (columnDesc.getTypeInfo().getCategory() == ObjectInspector.Category.PRIMITIVE
+                && ((PrimitiveTypeInfo) columnDesc.getTypeInfo()).getPrimitiveCategory() == PrimitiveCategory.TIMESTAMP) {
+          if (timestampPos != -1) {
+            throw new SemanticException("Multiple columns with timestamp type on query result; "
+                    + "could not resolve which one is the timestamp column");
+          }
+          timestampPos = i;
+        }
+      }
+      if (timestampPos == -1) {
+        throw new SemanticException("No column with timestamp type on query result; "
+                + "one column should be of timestamp type");
+      }
+      RowSchema selRS = new RowSchema(fsParent.getSchema());
+      // Granularity (partition) column
+      String udfName;
+      Class<? extends UDF> udfClass;
+      switch (parseCtx.getConf().getVar(HiveConf.ConfVars.HIVE_DRUID_INDEXING_GRANULARITY)) {
```

Comment:
This might not come necessarily from the HiveConf property anymore, since you included the possibility to override it with a table property. Thus, we need to check whether the Table associated with the FileSinkOperator contains the property and take it; otherwise use the value defined in Hive configuration as it is done here.